### PR TITLE
Add controller tests and CI

### PR DESCRIPTION
**Pull request by codablock**
This PR does multiple things:

1. Add some initial controller tests for ObjectTemplate. More will come via additional PRs.
2. Add workflows to run tests and multiple other checks.

### CODE REVIEW

File: controllers/suite_test.go [hunk 1]
```diff
@@ -37,21 +46,28 @@ import (
...
+var (
+	cfg       *rest.Config
+	k8sClient client.Client // You'll be using this client in your tests.
+	testEnv   *envtest.Environment
+	ctx       context.Context
+	cancel    context.CancelFunc
 
-func TestAPIs(t *testing.T) {
+	kubeconfigName = "template-controller-tests.kubeconfig"
+	kubeconfigPath string
```
Comment (erikgb) [inline controllers/suite_test.go:57]: What is the kubeconfig used for? Do we need it?

File: controllers/objecttemplate_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func buildTestConfigMap(name string, namespace string, data map[string]string) *unstructured.Unstructured {
+	m := map[string]any{}
+	for k, v := range data {
+		m[k] = v
+	}
+	return &unstructured.Unstructured{
```
Comment (erikgb) [inline controllers/objecttemplate_controller_test.go:39]: Why Unstructured? I prefer type-safety.

File: controllers/objecttemplate_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+		},
+	}
+}
+
+func buildTestSecret(name string, namespace string, data map[string]string) *unstructured.Unstructured {
+	m := map[string]any{}
+	for k, v := range data {
+		m[k] = v
+	}
+	return &unstructured.Unstructured{
```
Comment (erikgb) [inline controllers/objecttemplate_controller_test.go:57]: Why Unstructured? I prefer type-safety.

**Review by erikgb — COMMENTED** [review]
Very nice @codablock! This is a huge improvement! I've added a few comments/suggestions, but mostly in the nits category.

File: controllers/objecttemplate_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+package controllers
+
+import (
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"math/rand/v2"
+	client2 "sigs.k8s.io/controller-runtime/pkg/client"
```
[the remark this answers is not in the document]
Comment (codablock) [reply controllers/objecttemplate_controller_test.go:25]: Was not needed actually. For some reason gingko caused naming conflicts in the past due to the local import. But I really can't remember the details.

File: controllers/suite_test.go (hunk 1 above)
Comment (codablock) [reply controllers/suite_test.go:57]: I used it to debug RBAC issues, but I can actually remove it again

File: controllers/objecttemplate_controller_test.go (hunk 2 above)
Comment (codablock) [reply controllers/objecttemplate_controller_test.go:39]: Good point. Using typed objects now.

Comment (codablock) [reply controllers/objecttemplate_controller_test.go:39]: Ok, turns out the reason was simple: ObjectTemplate's template entries work with unstructured objects and there is no way around this. I'll at least use typed objects now and then convert them to unstructured.

**Comment by codablock** [comment]
Thanks @erikgb :) I fixed the issues you found and pushed it.

**Review by erikgb — APPROVED** [review]
LGTM!

**Pull request merged**